File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(socket: socket.socket, size):
    data = b""

    while len(data) < size:
        chunk = socket.recv(size - len(data))

        if not chunk:
            return b""

        data += chunk

    return data


def send_all(socket: socket.socket, bytes):
    sent = 0

    while sent < len(bytes):
        sent += socket.send(bytes[sent:])

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	for sent := 0; sent < len(bytes); {
		n, err := socket.Write(bytes[sent:])

		if err != nil {
			return err
		}
		sent += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buffer := make([]byte, size)

	for recv := 0; recv < size; {
		n, err := socket.Read(buffer[recv:])

		if err != nil {
			return buffer[:recv], err
		}
		recv += n
	}
	return buffer, nil
}
